### ISSUE TITLE
DynamicModuleLoader's children mounts twice

### ISSUE THREAD

**Issue by anilanar** [Microsoft/redux-dynamic-modules#66]
Due to the following workaround:

https://github.com/Microsoft/redux-dynamic-modules/blob/e1046942f26db108f7dbb6ccc593824c9de716f5/packages/redux-dynamic-modules/src/DynamicModuleLoader.tsx#L142-L147

children of `DynamicModuleLoader` is mounted twice, once with a `<Provider>` wrapper and once without the wrapper.

~I think this workaround is not needed anymore since React 16.6.3 and `store.getState()` gives the correct state on first render.~

Nevermind, it still behaves the same. I forked and unconditionally wrapped the component with the Provider always, which is better than double mounting IMO and has no side effects AFAIK.

**Comment by navneet-g** [comment]
@anilanar  thanks for reporting, Can you give a codesandbox with repro?

**Comment by anilanar** [comment]
https://codesandbox.io/s/8kx2jr7762

**Comment by navneet-g** [comment]
Sorry I didn't get a chance to look at it yet. I will take a look this week and get back to you.

**Comment by navneet-g** [comment]
@anilanar  I am not able to open the sandbox, I get following error. I will try again later today
![image](https://user-images.githubusercontent.com/13487927/54362135-a5bef400-4625-11e9-981a-6bd1df3d7f4f.png)

**Comment by anilanar** [comment]
Perhaps codesandbox had problems when you tried it, I can open it.

**Comment by navneet-g** [comment]
I can repro this and this is what I see, the storeState changes  and thus React rerenders children 
We need to debug more to see why the storeState is not upto date in first render after the module is loaded. Most probabbly this is another side effect of React's new context API, but for sure this is not a bug in ```redux-dynamic-modules```. It takes a long time to debug react, I will give it a try later in the week if I get sometime. 

![image](https://user-images.githubusercontent.com/13487927/54374615-bd55a700-463c-11e9-9bda-2a88e1a8d86b.png)

**Comment by shockk** [comment]
I have this trouble too. This issue blocks releasing our app (

**Comment by anilanar** [comment]
@navneet-g State not being up to date after updating a store is not a bug in `redux-dynamic-modules`. But using two different component roots (to overcome aforementioned problem) is a bug in `redux-dynamic-modules`.

**Comment by navneet-g** [comment]
@shockk  in our app we are using following version of react-redux, and redux and don't see this issue. We are using it with React 16.8. Please try these versions for now to unblock the release.
react-redux": "^5.1.1",   
    "redux": "^4.0.0",

**Comment by navneet-g** [comment]
@anilanar which two different component roots are you mentioning? Please elaborate.

**Comment by anilanar** [comment]
First the following:

https://github.com/Microsoft/redux-dynamic-modules/blob/master/packages/redux-dynamic-modules/src/DynamicModuleLoader.tsx#L135-L148

and 

https://github.com/Microsoft/redux-dynamic-modules/blob/master/packages/redux-dynamic-modules/src/DynamicModuleLoader.tsx#L116-L125

`_getLatestState` will be true on the first render and will be false on the second render. Thus, on the first render, it will render something like:

```
<ReactReduxContext.Provider ...>
  {children}
</ReactReduxContext.Provider>
```

On the second render, it will just render the `{children}`. Because the root of the component changes from `ReactReduxContext.Provider` to `children`'s root component, whatever is inside `{children}` will mount twice no matter what.

**Comment by navneet-g** [comment]
Thanks @anilanar ... @abettadapur can you take a look?

**Comment by abettadapur** [comment]
@anilanar @shockk  Sorry for the delay here. Could you guys try this version of the package and let me know if you still have issues?

`redux-dynamic-modules@3.5.0-alpha.3`

**Comment by anilanar** [comment]
@abettadapur It works. To understand the effects of `strictMode`, I created the following sandbox with bunch of console.logs, checking state update order of connected components inside and outside `DynamicModuleLoader` and also assigning `ref`s to `DynamicModuleLoader`'s children from outside.

https://codesandbox.io/s/73r20x4jzj

The difference seems to be only when mounting `DynamicModuleLoader`. `strictMode` changes the order of state updates. When it's turned off, `DynamicModuleLoader`s children's connected state updates before anyone else.

**Comment by abettadapur** [comment]
correct, the only difference should be at mount time. I don't think we can get around this for `strict` consumers.  Thanks for verifying and the example!

@navneet-g Let's move to publish the new version

**Comment by navneet-g** [comment]
Thanks @anilanar  and @abettadapur .
@abettadapur  I left a comment on your PR, let me know you want to address before merging or in a separate PR.
